core/embed/rust/src/ui/component/text/paragraphs.rs

```diff
@@ -116,30 +123,116 @@ where
+///
+/// The boxes are simply stacked below each other and may be further arranged
+/// before drawing.
+impl<'a, T> Iterator for PageBreakIterator<'a, T>
 where
     T: AsRef<[u8]>,
 {
-    type Msg = Never;
+    /// Paragraph index, character index, number of paragraphs shown.
+    type Item = (usize, usize, usize);
 
-    fn event(&mut self, _ctx: &mut EventCtx, _event: Event) -> Option<Self::Msg> {
-        None
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.para_offset >= self.paragraphs.list.len() {
+            return None;
+        }
+
+        let old_para_offset = self.para_offset;
+        let old_char_offset = self.char_offset;
+        let mut area = self.paragraphs.area;
+
+        for paragraph in self.paragraphs.list.iter_mut().skip(self.para_offset) {
+            loop {
+                paragraph.set_area(area);
+                let fit = paragraph.layout.layout_text(
+                    &paragraph.content.as_ref()[self.char_offset..],
+                    &mut paragraph.layout.initial_cursor(),
+                    &mut TextNoOp,
+                );
+                match fit {
+                    LayoutFit::Fitting { size, .. } => {
+                        // Text fits, update the bounding box.
+                        let (used, free) = area.hsplit(size.y);
+                        paragraph.set_area(used);
+                        // Continue with next paragraph in remaining space.
+                        area = free;
+                        self.char_offset = 0;
+                        self.para_offset += 1;
+                        break;
+                    }
+                    LayoutFit::OutOfBounds { processed_chars } => {
+                        // Text does not fit, assume whatever fits takes the entire remaining area.
+                        self.char_offset += processed_chars;
+                        let visible = if processed_chars > 0 {
+                            self.para_offset - old_para_offset + 1
+                        } else {
+                            self.para_offset - old_para_offset
+                        };
+                        // Return pointer to start of page.
+                        return Some((old_para_offset, old_char_offset, visible));
+                    }
+                }
+            }
+        }
+
+        // Last page.
+        Some((
+            old_para_offset,
+            old_char_offset,
+            self.para_offset - old_para_offset,
+        ))
     }
+}
 
-    fn paint(&mut self) {
-        self.layout.layout_text(
-            self.content.as_ref(),
-            &mut self.layout.initial_cursor(),
-            &mut TextRenderer,
-        );
+impl<T> Paginate for Paragraphs<T>
+where
+    T: AsRef<[u8]>,
+{
+    fn page_count(&mut self) -> usize {
+        // There's always at least one page.
+        let page_count = self.break_pages().count().max(1);
+
+        // Reset to first page.
+        self.change_page(0);
```

Comment:
why do you reset to first page as a side-effect?

Reply:
The `break_pages()` iterator modifies internal state (positions) of the individual `Paragraph`s so after counting pages it is inconsistent with the notion of current page that `Paragraphs` has. Resetting to first page at least makes these states consistent.

Also `change_page()` needs to be called before first `paint()` and `page_count()` is currently called right after constructing the object. Now that I write it out it sounds kinda stupid, will consider doing it differently.

Reply:
In 68a7be9 I rewrote `PageBreakIterator` to not mutate `Paragraph` state so reset is no longer needed, please take a look. `change_page(0)` must now be called by whoever constructed the `Paragraphs`.

The commit is in the follow-up PR #2071 because several commits there already modify this file. I can backport it here if you want.

Reply:
Further simplified (I hope) in faeb4dc